### ISSUE TITLE
Still not PKCE compliant as per RFC-7636

### ISSUE THREAD

**Issue by psibean** [Reinforz/twitter-oauth2-blog#3]
> This issue has been addressed in #2

That doesn't really fix the problem.

The code verifier should not be a hardcoded value, it's supposed to be unique for each request...

These values **are not supposed to be hardcoded at all** - they should be generated at runtime as per the PKCE standard [RFC-7636](https://datatracker.ietf.org/doc/html/rfc7636#section-4.1) sections 4.1 and 4.2.

    "We are hard coding `code_challenge` and `code_verifier` for simplicity. You can randomly generate it."

It's not a "can" it's a "should", hardcoding these values literally means you are not compliant with the PKCE workflow - it should NEVER be hardcoded, and that should be **made extremely clear**. since this is an educational context, you should educate by **explaining why this is a problem, and what people should be doing, and how that solves the problem.** Ideally you should explain what PKCE is, how it works, and the problems it solves. Or at least link to other resources that provide that information.

The whole point of PKCE is to provide CSRF protection on the authorization, using the same value each time **completely destroys the point**. you're making the security PKCE brings completely redundant without explaining so.  If you're creating an example for educational purposes, then you should be ensuring you are teaching and recommending best practices, or don't bother at all.

Also, the article hasn't been updated either, which is the main "educational resource" which references the repository. Providing people with resources which are not entirely informative over the security implications is increasingly problematic and is creating a circular influx of these kinds of sources.

>We are sending the access token in the cookie for simplicity. For a web application, we should store it somewhere more secure, like a database.

As previously mentioned, ideally you shouldn't be storing it in a database at all (and in some cases, depending on the provider, this may actually be against their terms and conditions). What you should be doing is **revoking the token** once you're done getting what you want from it. And that's because, OAuth2 is for authorisation, not for authentication.  This repo portrays itself for educational purposes, but the lack of explanation, reasoning, and the incorrect information, decreases the reliability and credibility.

**Comment by psibean** [comment]
I'm happy to solve this and update with a PR, it'll just take me some time to get to it, but I think it is important people are at least aware, in the mean time, but this example is not adequate, even for educational purposes, in it's current state.

**Comment by imoxto** [comment]
Of course, we would greatly appreciate a PR anytime! I will update the blog shortly with the current changes to the code challenge method, notes, etc.

**Comment by psibean** [comment]
If i get the time, I'll definitely put something up.

I do apologize for coming on quite aggressive, but this is a huge issue in the node / JS ecosystem. People consume a whole heap of tutorials which don't necessarily provide them with all of the insight or information, and then those people go off and continuously make mistakes based on the tutorials they followed, without even realizing it. Then they make their own tutorials with the same mistakes, and people rinse-and-repeat.

There is a **huge** amount of tutorials and "educational" resources out there which lack **extremely important** information, or simply contain **extremely dangerous and incorrect** examples. But it isn't necessarily their fault, they simply don't know any better, because they're following on from other people who did the exact same thing.

It's becoming an extremely frustrating thing for me that I keep seeing inadequate and incompetent resources/tutorials/examples that will lead to people referencing them and implementing extremely poor and insecure code.

**Comment by imoxto** [comment]
We greatly appreciate your constructive feedback. As this was the first article we published from our organization, there were a few mistakes, mostly due to a lack of in-depth knowledge (security aspect) and time constraint. We will try our best to minimize such mistakes in the future.
